### DISCUSSION TITLE
Integrate with R2R for scaleable / feature rich local RAG

### DISCUSSION THREAD

**Discussion by emrgnt-cmplxty** [open-webui/open-webui#3687] [Ideas]
**Is your feature request related to a problem? Please describe.**

open-webui offers an impressive suite of features, particularly its user-friendly interface. R2R has developed a powerful, flexible RAG (Retrieval-Augmented Generation) backend ([R2R docs](https://r2r-docs.sciphi.ai/introduction)). Integrating R2R as a RAG provider option in open-webui would:

1. Offer open-webui users more advanced RAG capabilities for complex use cases.
2. Provide R2R users access to a polished chat UI without building from scratch.

**Describe the solution you'd like**

Integrate R2R as an optional RAG provider within open-webui:

1. Add R2R as a selectable RAG backend option.
2. Implement API connections between open-webui's frontend and R2R's backend.
3. Expose R2R's advanced features through open-webui's interface.

As R2R core developers, we're eager to collaborate on this integration, providing expertise and implementation assistance.

**Describe alternatives you've considered**

1. Developing separate UIs: Leads to duplication of effort.
2. Building a plugin system: Requires extensive changes to open-webui's architecture.
3. Forking open-webui: Creates maintenance challenges and fragments the community.

Direct integration offers the best balance of leveraging existing work and providing value to both projects' users.

**Additional context**

This integration aligns with both projects' goals:

- For open-webui: Expands capabilities, offering users more choice in RAG implementations.
- For R2R: Provides access to a polished chat UI, allowing focus on core RAG technology.

This collaboration can exemplify open-source projects working together to create more robust AI tools.

Disclaimer: As an R2R core developer, I'm seeking project growth opportunities. This proposal is made in good faith, believing it will benefit both projects and their users.

We're open to discussing any questions or concerns the open-webui community might have.

**Comment by tjbck** [comment]
Great news, Open WebUI allows custom RAG pipeline using our new functions filter! I’ll add a few examples later today here, so we’d love for you to contribute!

**Reply by emrgnt-cmplxty** [reply]
that is great - looking forward to seeing that!

**Comment by emrgnt-cmplxty** [comment]
@tjbck - Could you share documentation?

**Reply by tjbck** [reply]
Busier than I thought atm, I'll do by best to share by eow this week, I appreciate your patience!🙌🏻

**Comment by tjbck** [comment]
@emrgnt-cmplxty Here's an example function for illustration purposes, Hope this helps! Feel free to join our discord server and drop any questions regarding functions if you encounter any roadblocks, I'd be happy to assist!

```py
"""
title: Custom RAG Filter Function
author: open-webui
author_url: https://github.com/open-webui
funding_url: https://github.com/open-webui
version: 0.1-rc1
required_open_webui_version: 0.3.8
"""

from pydantic import BaseModel, Field
from typing import Optional
import requests  # Importing the requests library to handle HTTP requests
import base64

from utils.misc import (
    get_last_user_message,
    add_or_update_system_message,
)


class Filter:
    class Valves(BaseModel):
        # Template for modifying context information in the user interface
        RAG_TEMPLATE: str = Field(
            default="""Use the following context as your learned knowledge, inside <context></context> XML tags.
<context>
    {{CONTEXT}}
</context>

When answering to user:
- If you don't know, just say that you don't know.
- If you're unsure, ask for clarification.
Avoid mentioning that you obtained the information from the context.
And answer according to the language of the user's question.

Given the context information, answer the user query.""",
            description="Template for system messages that embed contextual knowledge.",
        )

    def __init__(self):
        self.file_handler = True  # Enable custom file handling logic
        self.valves = self.Valves()

    def inlet(self, body: dict, __user__: Optional[dict] = None) -> dict:
        files = body.get("files", [])
        messages = body.get("messages", [])
        user_query = get_last_user_message(messages)
        context = None

        if files:
            files_payload = []
            try:
                for file in files:
                    file_path = file["file"]["meta"]["path"]
                    file_name = file["file"]["meta"]["name"]
                    print(f"Processing file from path: {file_path}")

                    # Opening file for sending; files must be closed by the requests library
                    file_handle = open(file_path, "rb")
                    files_payload.append(("files", (file_name, file_handle)))

                # Placeholder for server request to an external RAG (Retrieval-Augmented Generation) server
                # Example POST request:
                # r = requests.post(
                #     'https://external-rag-server.com/process',
                #     files=files_payload,
                #     data={'query': user_query}
                # )
                # r.raise_for_status()
                # res = r.json()
                # context = res.get('context', '')

                # For illustration without a real server:
                context = "Example context fetched based on the files and user query."
            except Exception as e:
                print(f"Request failed: {e}")
                # Handle request failure (You can add more sophisticated error handling)
                context = str(e)
            finally:
                # Ensure all file handles are closed
                for _, file_tuple in files_payload:
                    _, file_handle = file_tuple
                    file_handle.close()

            if context:
                # Related system message with retrieved context is injected into the message body
                system_message = self.valves.RAG_TEMPLATE.replace(
                    "{{CONTEXT}}", context
                )
                body["messages"] = add_or_update_system_message(
                    system_message, messages
                )
        return body
```

**Comment by gigq** [comment]
@emrgnt-cmplxty Any update on this, was looking for a better RAG solution and R2R seems like a good fit.

**Reply by emrgnt-cmplxty** [reply]
We'd be interested! It wasn't clear from me how to go about making the implementation, have the docs around this expanded?

**Reply by gigq** [reply]
Yeah I'm not sure, I'm just a user hoping for better rag in open webui. It seems like in June or July there was some traction on issue #1293 but it has stalled now so wasn't sure if I was just missing some work happening behind the scenes. It sounds like no though.

**Reply by jannikstdl** [reply]
Hi, I think this all comes together with #3527 in version 0.4.x

But for the R2R integration I'm not sure but it should work with the function provided by tim.
